bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java

```diff
@@ -130,27 +132,100 @@ public void run()
         storageHelperExecutor.shutdown();
         try {
           serverHelperExecutor.awaitTermination(5000, TimeUnit.MILLISECONDS);
+          storageHelperExecutor.awaitTermination(5000, TimeUnit.MILLISECONDS);
         } catch (InterruptedException ex) {
           logger.debug("Executor Termination", ex);
         }
-        logger.info("Server stopped listening at {}", address);
+        logger.info("Server stopped listening at {}", getServerAddress());
+        latch.countDown();
+        latch = null;
```

Comment:
It is used as a gate. From CountDownLatch java doc: "A CountDownLatch is a versatile synchronization tool and can be used for a number of purposes. A CountDownLatch initialized with a count of one serves as a simple on/off latch, or gate: all threads invoking await wait at the gate until it is opened by a thread invoking countDown()." I don't think it is worth re-implementing the gate functionality of the CountDownLatch class using simple object wait, especially that the class is provided by JDK and not a 3rd party library.